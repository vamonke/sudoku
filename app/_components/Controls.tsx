type Props = {
  restart: () => void;
  newGame: () => void;
  undo: () => void;
  toggleShowConflicts: () => void;
  showConflicts: boolean;
};

export default function Controls(props: Props) {
  const { restart, newGame, undo, toggleShowConflicts, showConflicts } = props;

  return (
    <div className="flex flex-row gap-4 mt-2 justify-center items-center">
      <button onClick={restart}>Restart (R)</button>
      <button onClick={newGame}>New Game (N)</button>
      <button onClick={undo}>Undo (Z)</button>
      <button onClick={toggleShowConflicts}>
        {showConflicts ? "Hide Hints (H)" : "Show Hints (H)"}
      </button>
    </div>
  );
}
