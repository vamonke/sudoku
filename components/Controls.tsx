import { useEffect } from "react";

type Props = {
  reset: () => void;
  newGame: () => void;
  undo: () => void;
  toggleShowConflict: () => void;
  showConflict: boolean;
};

export default function Controls(props: Props) {
  const { reset, newGame, undo, toggleShowConflict, showConflict } = props;

  useEffect(() => {
    const eventListener = (event: KeyboardEvent) => {
      if (event.key === "r") reset(); // TODO: Add confirmation
      if (event.key === "n") newGame(); // TODO: Add confirmation
      if (event.key === "w") toggleShowConflict();
      if (event.key === "z") undo();
    };
    window.addEventListener("keyup", eventListener);
    return () => {
      window.removeEventListener("keyup", eventListener);
    };
  }, [newGame, reset, toggleShowConflict, undo]);

  return (
    <div className="flex flex-row gap-4 mt-2 justify-center items-center">
      <button data-testid="reset" onClick={reset}>
        Reset (R)
      </button>
      <button data-testid="new" onClick={newGame}>
        New Game (N)
      </button>
      <button data-testid="undo" onClick={undo}>
        Undo (Z)
      </button>
      <button data-testid="toggle" onClick={toggleShowConflict}>
        {showConflict ? "Hide warnings (W)" : "Show warnings (W)"}
      </button>
    </div>
  );
}
