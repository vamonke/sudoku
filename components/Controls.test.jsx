import "@testing-library/jest-dom";
import { render, fireEvent, screen } from "@testing-library/react";
import Controls from "./Controls";

describe("Controls", () => {
  describe("renders correctly", () => {
    it("when showConflict is true", () => {
      const { container } = render(<Controls showConflict={true} />);
      expect(container).toMatchSnapshot();
    });

    it("when showConflict is false", () => {
      const { container } = render(<Controls showConflict={false} />);
      expect(container).toMatchSnapshot();
    });
  });

  describe("reset function", () => {
    it("is called when the reset button is clicked", () => {
      const mockReset = jest.fn();
      render(<Controls reset={mockReset} />);
      const resetButton = screen.getByTestId("reset");
      resetButton.click();
      expect(mockReset).toHaveBeenCalled();
    });

    it("is called when r key is pressed", () => {
      const mockReset = jest.fn();
      const { container } = render(<Controls reset={mockReset} />);
      fireEvent.keyUp(container, { key: "r" });
      expect(mockReset).toHaveBeenCalled();
    });
  });

  describe("newGame", () => {
    it("is called when the new button is clicked", () => {
      const mockNewGame = jest.fn();
      render(<Controls newGame={mockNewGame} />);
      const newButton = screen.getByTestId("new");
      newButton.click();
      expect(mockNewGame).toHaveBeenCalled();
    });

    it("is called when n key is pressed", () => {
      const mockNewGame = jest.fn();
      const { container } = render(<Controls newGame={mockNewGame} />);
      fireEvent.keyUp(container, { key: "n" });
      expect(mockNewGame).toHaveBeenCalled();
    });
  });

  describe("undo function", () => {
    it("is called when the undo button is clicked", () => {
      const mockUndo = jest.fn();
      render(<Controls undo={mockUndo} />);
      const undoButton = screen.getByTestId("undo");
      undoButton.click();
      expect(mockUndo).toHaveBeenCalled();
    });

    it("is called when z key is pressed", () => {
      const mockUndo = jest.fn();
      const { container } = render(<Controls undo={mockUndo} />);
      fireEvent.keyUp(container, { key: "z" });
      expect(mockUndo).toHaveBeenCalled();
    });
  });

  describe("toggleShowConflict function", () => {
    it("is called when the toggle button is clicked", () => {
      const mockToggleShowConflict = jest.fn();
      render(<Controls toggleShowConflict={mockToggleShowConflict} />);
      const toggleButton = screen.getByTestId("toggle");
      toggleButton.click();
      expect(mockToggleShowConflict).toHaveBeenCalled();
    });

    it("is called when w key is pressed", () => {
      const mockToggleShowConflict = jest.fn();
      const { container } = render(
        <Controls toggleShowConflict={mockToggleShowConflict} />
      );
      fireEvent.keyUp(container, { key: "w" });
      expect(mockToggleShowConflict).toHaveBeenCalled();
    });
  });
});
