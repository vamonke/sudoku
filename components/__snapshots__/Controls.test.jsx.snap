// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls renders correctly when showConflict is false 1`] = `
<div>
  <div
    class="flex flex-row gap-4 mt-2 justify-center items-center"
  >
    <button
      data-testid="reset"
    >
      Reset (R)
    </button>
    <button
      data-testid="new"
    >
      New Game (N)
    </button>
    <button
      data-testid="undo"
    >
      Undo (Z)
    </button>
    <button
      data-testid="toggle"
    >
      Show warnings (W)
    </button>
  </div>
</div>
`;

exports[`Controls renders correctly when showConflict is true 1`] = `
<div>
  <div
    class="flex flex-row gap-4 mt-2 justify-center items-center"
  >
    <button
      data-testid="reset"
    >
      Reset (R)
    </button>
    <button
      data-testid="new"
    >
      New Game (N)
    </button>
    <button
      data-testid="undo"
    >
      Undo (Z)
    </button>
    <button
      data-testid="toggle"
    >
      Hide warnings (W)
    </button>
  </div>
</div>
`;
